File: Solutions/MoNaDDu/Game Shop/3/Part3.hs
module Part2 where

import Control.Monad

data Item = Item { name :: String, description :: String, weight :: Int, value :: Int } |
            Weapon { name :: String, description :: String, weight :: Int, value :: Int, damage :: Int } |
            Armor { name :: String, description :: String, weight :: Int, value :: Int, defense :: Int } |
            Potion { name :: String, description :: String, weight :: Int, value :: Int, potionType :: PotionType, capacity :: Int }
 deriving ( Show )

data PotionType = HP | MP

instance Show PotionType where
    show HP = "Health"
    show MP = "Mana"

data Shop = Shop { shopName :: String, shopItems :: Items }

type Items = [Item]

main :: IO ()
main = do
    printMenu
    input <- getLine
    play $ read input
    Control.Monad.unless (read input == 3) main

printMenu :: IO ()
printMenu = do
    putStrLn "- Shop Select -"
    putStrLn "   1. Weapon/Armor Shop"
    putStrLn "   2. Potion Shop"
    putStrLn "   3. Exit\n"
    putStr "Select : "
    return ()

play :: Int -> IO ()
play n
    | n == 1 = showItemList weaponAndArmorShop
    | n == 2 = showItemList potionShop
    | n == 3 = return ()
    | otherwise = putStrLn "\nInvalid number! Try again.\n"



testCase1 = describe items

describe :: Items -> IO ()
describe = mapM_ (putStrLn . toString)

toString :: Item -> String
toString (Item name description weight value) = baseInfo (Item name description weight value)
toString (Weapon name description weight value damage) = baseInfo (Item name description weight value) ++
                                                         "Damage      = " ++ show damage ++ "\n"
toString (Armor name description weight value defense) = baseInfo (Item name description weight value) ++
                                                         "Defense     = " ++ show defense ++ "\n"
toString (Potion name description weight value potionType capacity) = baseInfo (Item name description weight value) ++
                                                                      "Type         = " ++ show potionType ++ "\n" ++
                                                                      "Capacity     = " ++ show capacity ++ "\n"

baseInfo :: Item -> String
baseInfo item = "Name        = " ++ name item ++ "\n" ++
                "Description = " ++ description item ++ "\n" ++
                "Weight      = " ++ w'' ++ "\n" ++
                "value       = " ++ v'' ++ "\n"
                where 
                w = weight item
                v = value item
                w' = show w
                v' = show v
                w''
                 | w > 1 = w' ++ " lbs"
                 | otherwise = w' ++ " lb"
                v''
                 | v > 1 = v' ++ " gold coins"
                 | otherwise = v' ++ " gold coin"

showItemList :: Shop -> IO ()
showItemList shop = do
    putStrLn $ "Welcom to " ++ shopName shop ++ "!\n"
    describe $ shopItems shop
    putStrLn "\n"

items = [
    Weapon { name = "Excalibur", description = "The legendary sword of King Arthur", weight = 12, value = 1024, damage = 24 },
    Armor { name = "Steel Armor", description = "Protective covering made by steel", weight = 15, value = 805, defense = 18}
    ]

weaponAndArmorShop = Shop { shopName = "Weapon/Armor", shopItems = [
    Weapon { name = "Sword", description = "Medium DMG", weight = 3, value = 10, damage = 10 },
    Armor { name = "Cap", description = "Light Armor", weight = 1, value = 5, defense = 5 },
    Armor { name = "Gloves", description = "Light Armor", weight = 1, value = 5, defense = 5 },
    Weapon { name = "Axe", description = "High Dmg", weight = 5, value = 15, damage = 15 },
    Armor { name = "Boots", description = "Light Armor", weight = 1, value = 5, defense = 5 }
    ] }

potionShop = Shop { shopName = "Potion", shopItems = [
    Potion { name = "Small Health Potion", description = "Recovery 100 HP", weight = 2, value = 5, potionType = HP, capacity = 100 },
    Potion { name = "Small Mana Potion", description = "Recovery 50 MP", weight = 1, value = 30, potionType = MP, capacity = 50 },
    Potion { name = "Medium Health Potion", description = "Recovery 200 HP", weight = 4, value = 120, potionType = HP, capacity = 200 },
    Potion { name = "Medium Mana Potion", description = "Recovery 100 MP", weight = 2, value = 75, potionType = MP, capacity = 100 },
    Potion { name = "Large Health Potion", description = "Recovery 300 HP", weight = 6, value = 200, potionType = HP, capacity = 300 }
    ] }

File: Solutions/MoNaDDu/Game Shop/3/Part3.cpp
#include <iostream>
#include <string>
#include <iomanip>
#include <vector>
#include <memory>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Item;
class Weapon;
class Armor;
class Potion;
class Shop;
class ShopManager;

using ItemPtr = shared_ptr<Item>;
using WeaponPtr = shared_ptr<Weapon>;
using ArmorPtr = shared_ptr<Armor>;
using PotionPtr = shared_ptr<Potion>;
using ShopPtr = shared_ptr<Shop>;
using ShopManagerPtr = shared_ptr<ShopManager>;

class Item
{
public:
	Item(const string& name, const string& description, const int weight, const int value)
		: name(name), description(description), weight(weight), value(value)
	{

	}

	virtual ~Item() = default;

	virtual void Describe(void) const
	{
		DescribeName();
		DescribeDescription();
		DescribeWeight();
		DescribeValue();
	}

protected:
	static int GetSpaceInfo(void)
	{
		return space;
	}

private:
	string name;
	string description;
	int weight;
	int value;
	static const int space = 12;

	inline void DescribeName(void) const
	{
		cout << left << setw(space) << "Name" << "= " << name << endl;
	}

	inline void DescribeDescription(void) const
	{
		cout << left << setw(space) << "Description" << "= " << description << endl;
	}

	inline void DescribeWeight(void) const
	{
		cout << left << setw(space) << "Weight" << "= " << weight << (weight > 1 ? " lbs" : " lb") << endl;
	}

	inline void DescribeValue(void) const
	{
		cout << left << setw(space) << "Value" << "= " << value << (value > 1 ? " gold coin" : " gold coins") << endl;
	}
};

class Weapon : public Item
{
public:
	static ItemPtr MakeInstance(const string& name, const string& description, const int weight, const int value, const int damage)
	{
		return make_shared<Weapon>(name, description, weight, value, damage);
	}

	Weapon(const string& name, const string& description, const int weight, const int value, const int damage)
		: Item(name, description, weight, value), damage(damage)
	{

	}

	virtual ~Weapon() = default;

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeDamage();
	}

private:
	int damage;

	inline void DescribeDamage(void) const
	{
		cout << setw(GetSpaceInfo()) << "Damage" << "= " << damage << endl;
	}
};

class Armor : public Item
{
public:
	static ItemPtr MakeInstance(const string& name, const string& description, const int weight, const int value, const int defense)
	{
		return make_shared<Armor>(name, description, weight, value, defense);
	}

	Armor(const string& name, const string& description, const int weight, const int value, const int defense)
		: Item(name, description, weight, value), defense(defense)
	{

	}

	virtual ~Armor() = default;

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeDefense();
	}

private:
	int defense;

	inline void DescribeDefense(void) const
	{
		cout << setw(GetSpaceInfo()) << "Defense" << "= " << defense << endl;
	}
};

class Potion : public Item
{
public:
	enum class Type
	{
		HP,
		MP
	};

	static ItemPtr MakeInstance(const string& name, const string& description, const int weight, const int value, const Type type, const int capacity)
	{
		return make_shared<Potion>(name, description, weight, value, type, capacity);
	}

	Potion(const string& name, const string& description, const int weight, const int value, const Type type, const int capacity)
		: Item(name, description, weight, value), type(type), capacity(capacity)
	{

	}

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeType();
		DescribeCapacity();
	}

private:
	const int capacity;
	const Type type;
	static const unordered_map<Type, string> type_to_string;

	inline void DescribeType(void) const
	{
		cout << setw(GetSpaceInfo()) << "Type" << "= " << type_to_string.at(type) << endl;
	}

	inline void DescribeCapacity(void) const
	{
		cout << setw(GetSpaceInfo()) << "Capacity" << "= " << capacity << endl;
	}
};

const unordered_map<Potion::Type, string> Potion::type_to_string{
	make_pair(Potion::Type::HP, "Health"),
	make_pair(Potion::Type::MP, "Mana")
};

class Shop : public enable_shared_from_this<Shop>
{
public:
	static ShopPtr MakeInstance(const string& name = "", vector<ItemPtr> items = vector<ItemPtr>())
	{
		return make_shared<Shop>(name, items);
	}

	Shop() = default;

	Shop(const string& name = "", vector<ItemPtr> items = vector<ItemPtr>())
		: name(name), items(items)
	{

	}

	virtual ~Shop() = default;

	ShopPtr PushBack(ItemPtr item)
	{
		items.push_back(item);

		return shared_from_this();
	}

	void ShowItemList(void) const
	{
		cout << "Welcome to " << name << "!\n";
		cout << "- Item List -\n";
		for (const auto& item : items)
		{
			item->Describe();
			cout << endl;
		}
	}

	const string& GetName(void) const
	{
		return name;
	}

private:
	const string name;
	vector<ItemPtr> items;
};

class ShopManager : public enable_shared_from_this<ShopManager>
{
public:
	static ShopManagerPtr MakeInstance(void)
	{
		return make_shared<ShopManager>();
	}

	ShopManagerPtr PushBack(ShopPtr shop)
	{
		shop_list.push_back(shop);

		return shared_from_this();
	}

	template<typename F>
	void PrintShopName(F Func) const
	{
		for (const auto& name : shop_list)
			Func(name->GetName());
	}

	void ShowItemListOfShop(const size_t index)
	{
		shop_list[index]->ShowItemList();
	}

	size_t GetNumberOfShop(void) const
	{
		return shop_list.size();
	}

private:
	vector<ShopPtr> shop_list;
};

ShopManagerPtr InitShopManager(void)
{
	return ShopManager::MakeInstance()
		->PushBack(
			Shop::MakeInstance("Weapon/Armor", {
				Weapon::MakeInstance("Sword", "Medium DMG", 3, 10, 10),
				Armor::MakeInstance("Cap", "Light Armor", 1, 5, 5),
				Armor::MakeInstance("Gloves", "Light Armor", 1, 5, 5),
				Weapon::MakeInstance("Axe", "High DMG", 5, 15, 15),
				Armor::MakeInstance("Boots", "Light Armor", 1, 5, 5)
			})
		)
		->PushBack(
			Shop::MakeInstance("Potion", {
				Potion::MakeInstance("Small Health Potion", "Recovery 100 HP", 2, 5, Potion::Type::HP, 100),
				Potion::MakeInstance("Small Mana Potion", "Recovery 50 MP", 1, 30, Potion::Type::MP, 50),
				Potion::MakeInstance("Medium Health Potion", "Recovery 200 HP", 4, 120, Potion::Type::HP, 200),
				Potion::MakeInstance("Medium Mana Potion", "Recovery 100 MP", 2, 75, Potion::Type::MP, 100),
				Potion::MakeInstance("Large Health Potion", "Recovery 300 HP", 6, 200, Potion::Type::HP, 300)
			})
		);
}

int Play(const ShopManagerPtr& shop_manager);

void PrintMenu(const ShopManagerPtr& shop_manager)
{
	auto index = 0u;
	cout << " - Shop Select -\n";
	shop_manager->PrintShopName([&index](const auto& name) {
		cout << "    " << ++index << ". " << name << " Shop\n";
	});
	cout << "    " << ++index << ". Exit\n\n";
	cout << "Select : ";
}

size_t Input(void)
{
	size_t selected_number;

	cin >> selected_number;

	return selected_number - 1;
}

int Play(const ShopManagerPtr& shop_manager)
{
	PrintMenu(shop_manager);

	auto selected_number = Input();
	auto end = false;

	if (selected_number < 0 || selected_number > shop_manager->GetNumberOfShop())
		cout << "\nInvalid number! Try again.\n" << endl;
	else if (selected_number != shop_manager->GetNumberOfShop())
		shop_manager->ShowItemListOfShop(selected_number);
	else
		end = true;

	return end ? 0 : Play(shop_manager);
}

int main()
{
	auto shop_manager = InitShopManager();

    return Play(shop_manager);
}